### PR TITLE
service broker uses conjur version

### PR DESCRIPTION
**Pull request by jtuttle**
https://ca-il-jira.il.cyber-ark.com:8443/browse/CONJ-4692

### CODE REVIEW

File: app/models/service_binding.rb [hunk 1]
```diff
@@ -28,34 +28,55 @@ def create(app_id)
...
     host = conjur_api.role(role_name)
+    
     raise HostNotFound if !host.exists?
 
     host.rotate_api_key
-    load_policy template_delete, method: Conjur::API::POLICY_METHOD_PATCH
+
+    if ConjurClient.version == 5
+      load_policy template_delete, method: Conjur::API::POLICY_METHOD_PATCH
+    end
```
Comment (izgeri) [inline app/models/service_binding.rb]: are we still planning to `retire` the host? is there a separate story for that?

File: app/models/service_binding.rb [hunk 2]
```diff
@@ -28,34 +28,55 @@ def create(app_id)
...
+    host_id =
       if ConjurClient.policy != 'root'
-        "host/#{ConjurClient.policy}/#{@binding_id}"
+        "#{ConjurClient.policy}/#{@binding_id}"
       else
-        "host/#{@binding_id}"
+        @binding_id
       end
     
+    api_key = (ConjurClient.version == 4 ? create_v4(host_id) : create_v5)
```
Comment (izgeri) [inline app/models/service_binding.rb]: `create` to me doesn't imply that this method returns the api key after creating - can you update the name of these methods to make this clearer?

File: app/models/service_binding.rb [hunk 3]
```diff
@@ -28,34 +28,55 @@ def create(app_id)
...
+      load_policy template_delete, method: Conjur::API::POLICY_METHOD_PATCH
+    end
   end
 
   private
 
+  def create_v4(host_id)
+    hf_token =
+      conjur_api.
+        resource("cucumber:host_factory:pcf%2Fpcf-apps").
```
Comment (izgeri) [inline app/models/service_binding.rb]: the first `pcf` here should be `ConjurClient.policy`, and `cucumber` should be `ConjurClient.account`

File: app/models/service_binding.rb [hunk 4]
```diff
@@ -28,34 +28,55 @@ def create(app_id)
...
+        create_token(Time.now + 1.hour)
+
+    host = Conjur::API.host_factory_create_host(hf_token, host_id)
+
+    host.api_key
+  end
+
+  def create_v5
```
Comment (izgeri) [inline app/models/service_binding.rb]: I see why you don't pass the `host_id` here, bc you load the `binding_id` into the `ConjurClient.policy` policy namespace, but it seems worth noting something about for future engs who are in this code, bc this is subtle

File: ci/configure_v4.sh [hunk 5]
```diff
@@ -11,3 +11,4 @@ env CONJUR_AUTHN_LOGIN=admin CONJUR_AUTHN_API_KEY=secret conjur policy load --as
 CONFIGURE
 
 docker cp $(docker-compose ps -q conjur_4):/opt/conjur/etc/ssl/ca.pem ./tmp/conjur.pem
+export CONJUR_SSL_CERTIFICATE="$(cat tmp/conjur.pem)"
```
Comment (izgeri) [inline ci/configure_v4.sh]: this was moved to test.sh in my last PR

File: lib/conjur_client.rb [hunk 6]
```diff
@@ -4,7 +4,11 @@
 class ConjurClient
   class << self
     def api
-      @@api
+      ConjurClient.new.api
+    end
+
+    def version
```
Comment (izgeri) [inline lib/conjur_client.rb]: this is a duplicate

Comment (jtuttle) [reply lib/conjur_client.rb]: Thanks git merge!

File: ci/configure_v4.sh (hunk 5 above)
Comment (jtuttle) [reply ci/configure_v4.sh]: Removed.

File: app/models/service_binding.rb (hunk 3 above)
Comment (jtuttle) [reply app/models/service_binding.rb]: Oops, forgot to clean this up. Will do so.

File: app/models/service_binding.rb (hunk 4 above)
Comment (jtuttle) [reply app/models/service_binding.rb]: Changed things around to make the args consistent.

File: app/models/service_binding.rb (hunk 2 above)
Comment (jtuttle) [reply app/models/service_binding.rb]: Hmm, any suggestions for a better name? The method is creating a host so I'm not sure what could be better. I don't think it's totally bizarre to return an api key from a host creation operation but I'm open to ideas!

File: app/models/service_binding.rb (hunk 1 above)
Comment (jtuttle) [reply app/models/service_binding.rb]: I don't see anything about retiring hosts in the API v4 router. Will try to figure out what's going on there.

Comment (jtuttle) [reply app/models/service_binding.rb]: Let's punt on this for now. Can only be done through the CLI and don't think it's really necessary.

File: lib/conjur_client.rb [hunk 7]
```diff
@@ -4,7 +4,15 @@
 class ConjurClient
   class << self
     def api
-      @@api
+      ConjurClient.new.api
+    end
+
+    def v4_host_factory_id
+      "#{account}:host_factory:#{policy}/pcf-apps"
```
Comment (izgeri) [inline lib/conjur_client.rb]: ack. the second pcf is actually `#{policy}` too.

Comment (jtuttle) [reply lib/conjur_client.rb]: Fixed

**Comment by izgeri** [comment]
LGTM :)

**Pull request merged**